[Merged by Bors] - feat({algebra/group_action_hom, data/equiv/mul_add}): add missing `inverse` defs

ocfnash opened the pull request:
---


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/non_unital_alg_hom): define `non_unital_alg_hom`

The motivation is to be able to state the universal property
for a magma algebra using bundled morphisms.

eric-wieser commented on src/algebra/non_unital_alg_hom.lean:
```diff
@@ -0,0 +1,171 @@
+variables [comm_semiring R] [semiring A] [semiring B] [algebra R A] [algebra R B]
+
+instance non_unital_alg_hom.has_coe : has_coe (A →ₐ[R] B) (non_unital_alg_hom R A B) :=
+⟨λ f, { map_smul' := f.map_smul, .. f, }⟩
```
For consistency with all the other hom conversions, I'd probably lean towards having an explicit `to_non_unital_alg_hom` definition too.

ocfnash replied:
Agreed, done.

eric-wieser commented on src/algebra/non_unital_alg_hom.lean:
```diff
@@ -0,0 +1,171 @@
+structure non_unital_alg_hom [semiring R]
+  [non_unital_non_assoc_semiring A] [module R A]
+  [non_unital_non_assoc_semiring B] [module R B]
+  extends A →ₗ[R] B, mul_hom A B
```
In theory you could relax this to
```suggestion
structure non_unital_alg_hom [monoid R]
  [non_unital_non_assoc_semiring A] [distrib_mul_action R A]
  [non_unital_non_assoc_semiring B] [distrib_mul_action R B]
  extends A →+[R] B, mul_hom A B
```

which then lets you express things like `non_unital_alg_hom (units R) A B`

ocfnash replied:
Hmm, this is _probably_ a good idea. I've just added a commit which does it anyway so let's see if it builds etc.

ocfnash committed:
Add `to_non_unital_alg_hom`

ocfnash committed:
Generalise from `module` to just `distrib_mul_action`

ocfnash commented on src/algebra/group_action_hom.lean:
```diff
@@ -103,8 +115,22 @@ def to_quotient : G →[G] quotient_group.quotient H :=
 
+-- TODO Is this _really_ missing!!? Where should it go?
+/-- The inverse of a bijective `monoid_hom` is a `monoid_hom`. -/
+@[to_additive "The inverse of a bijective `add_monoid_hom` is an `add_monoid_hom`."]
```
Surely I'm missing the obvious. This _has_ to exist already but I couldn't find it!

eric-wieser commented on src/algebra/group_action_hom.lean:
```diff
@@ -103,8 +115,22 @@ def to_quotient : G →[G] quotient_group.quotient H :=
+    calc g 1 = g (f 1) : by rw f.map_one
+         ... = 1 : by rw h₁,
+  map_mul' := λ x y,
+    calc g (x * y) = g ((f (g x)) * (f (g y))) : by rw [h₂, h₂]
```
g.to_mul_hom.inverse will save you this part of the proof

ocfnash replied:
Thanks.

ocfnash committed:
Golf and move `monoid_hom.inverse`

eric-wieser commented on src/data/equiv/mul_add.lean:
```diff
@@ -45,6 +45,14 @@ def mul_hom.inverse [has_mul M] [has_mul N] (f : mul_hom M N) (g : N → M)
+  B →* A :=
+{ map_one' := show g 1 = 1, by rw [← f.map_one, h₁],
+  .. (f : mul_hom A B).inverse g h₁ h₂, }
+
```
In the interest of reducing build times, it might make sense to spin this change and the other new `inverse`s off into their own PR.

ocfnash replied:
I actually span out the new `non_unital_alg_hom` so we could reuse the oleans from this PR for the other changes.

ocfnash committed:
Drop addition of `non_unital_alg_hom.lean`

The plan is to split this PR in two since the other changes are
deep and cause long build times.

eric-wieser commented on src/algebra/group_action_hom.lean:
```diff
@@ -137,6 +150,24 @@ variables {M A B}
+
+instance : has_one (A →+[M] A) :=
+⟨{ map_smul' := by simp,
+   .. add_monoid_hom.id A }⟩
```
If you defined this further down the file you'd have `id` available

ocfnash replied:
OK I've added https://github.com/leanprover-community/mathlib/pull/7847/commits/1322b50f77d47cc82aab052da43a68cea08cd0df to use `id`.

ocfnash committed:
Define `distrib_mul_action_hom.has_one` via `id`.

eric-wieser commented on src/algebra/group_action_hom.lean:
```diff
@@ -162,6 +175,22 @@ protected def id : A →+[M] A :=
+
+@[simp] lemma coe_zero : ((0 : A →+[M] B) : A → B) = 0 := rfl
+
+@[simp] lemma coe_one : ((1 : A →+[M] A) : A → A) = id := rfl
```
I'm a little surprised you added these - they're certainly worth having eventually, but I'm surprised they're useful to you without also `has_add` (elementwise) and `has_mul` (composition), along with the associated monoid structures.

ocfnash replied:
You're quite right that I'm not using them but my view is that there's not much point in adding a `one` or `zero` if I don't characterise them.

ocfnash replied:
Oh I think you mean you're surprised that I added instances of `has_zero`, `has_one` rather than the `coe_zero`, `coe_one` lemmas given that I added the instances.

Again I don't need them but I figured that while I was here I might as well remove the `@[nolint has_inhabited_instance]` by adding zero as a default and then I figured if I was adding zero, why not add one!

eric-wieser replied:
Yes exactly - this seemed orthogonal to both the rest of this PR and your original, so I thought I'd check why it was here. Satisfying the linter was also my best guess.

eric-wieser commented on src/algebra/group_action_hom.lean:
```diff
@@ -93,6 +93,18 @@ ext $ λ x, by rw [comp_apply, id_apply]
+variables {A B}
+
+/-- The inverse of a bijective equivariant map is equivariant. -/
+def inverse (f : A →[M] B) (g : B → A)
```
```suggestion
@[simps]
def inverse (f : A →[M] B) (g : B → A)
```
and below

eric-wieser commented on src/data/equiv/mul_add.lean:
```diff
@@ -45,6 +45,14 @@ def mul_hom.inverse [has_mul M] [has_mul N] (f : mul_hom M N) (g : N → M)
+def monoid_hom.inverse {A B : Type*} [monoid A] [monoid B] (f : A →* B) (g : B → A)
+  (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f) :
+  B →* A :=
+{ map_one' := show g 1 = 1, by rw [← f.map_one, h₁],
```
```suggestion
{ to_fun := g,
  map_one' := by rw [← f.map_one, h₁],
```
First line needed to make simp happy, but may also allow removing the show

ocfnash replied:
I believe I understand: the `inverse_apply` lemma generated by `simps` will include a `.to_fun` without this?

IIUC this is not so serious because of `to_fun_eq_coe` but still an improvement, and anyway, it allows dropping the `show` as you mention.

ocfnash committed:
Add simps decorators

Vierkantor reviewed (approved):
Thanks!

bors r+

kbuzzard commented on src/algebra/group_action_hom.lean:
```diff
@@ -124,10 +135,12 @@ instance has_coe' : has_coe (A →+[M] B) (A →[M] B) :=
 
 instance : has_coe_to_fun (A →+[M] B) :=
-⟨_, λ c, c.to_fun⟩
+⟨_, to_fun⟩
```
There's another one of these on line 238
